spec.md

```diff
@@ -2232,6 +2331,58 @@ message VolumeUsage {
   // Units by which values are measured. This field is REQUIRED.
   Unit unit = 4;
 }
+
+message VolumeHealth {
+  option (alpha_message) = true;
+
+  enum Condition {
+    // Volume health condition is unknown; If this is the initial
+    // report of volume health, treat it as healthy. If there was
+    // already a condition reported previously, treat unknown as
+    // unchanged from the previous condition.
+    UNKNOWN = 0;
+    // Volume is accessible and no problems are detected. 
+    HEALTHY = 1;
+    // Volume is not accessible currently, but there's no reason to
+    // expect data loss. Examples: power outage, network outage,
+    // maintenance.
+    TEMPORARY_INACCESSIBLE = 2;
+    // Volume is accessible currently, but is degraded. Data loss
+    // is unlikely as the storage system is rebuilding to fix the
+    // problem automatically. For example, a replica is lost on
+    // the storage system but a spare is configured, and the storage
+    // system is resyncing/reconstructing.
+    TEMPORARY_DEGRADED = 3;
+    // Volume is accessible currently, but a problem is detected that
+    // could lead to eventual data loss. Examples: recoverable hardware
+    // failure (i.e. the disk needs to be manually replaced by a human),
+    // cooling failure, equipment significantly past end of life.
+    FAILURE_LIKELY = 4;
+    // Volume is not accessible currently, and there is reason to
+    // believe the outage may be permanent. Examples: unrecoverable
+    // hardware failure, natural disasters, damage to equipment.
+    FATAL = 5;
+  }
+
+  // The level of the health condition of the volume.
+  // This field is REQUIRED.
+  Condition condition = 1;
+
+  // The error code and message describing the health condition of
+  // the volume. This field is REQUIRED.
+  repeated VolumeHealthError error = 2;
+}
+
+message VolumeHealthError {
+  // The error code describing the health condition of the volume.
+  // This is an opaque field to CO.
+  // This field is REQUIRED.
+  string error_code = 1;
+
+  // The error message associated with the above error code.
+  // This field is OPTIONAL.
+  string message = 2;
```

Comment:
I think the OPTIONAL/REQUIRED here are backwards. Messages should be REQUIRED, and error codes should be OPTIONAL. It's all opaque and requires human eyes/brains to parse - a message seems like the minimum bar to clear here. Otherwise if admin Dan reads one of these serialized VolumeHealthError objects and sees just a code (w/o a message) that's way less informative than a meaningful error message (w/o an error code).

Reply:
Sure